## Use the following code for the questions below:

In [1]:
import numpy as np


In [2]:
a = np.array([4, 10, 12, 23, -2, -1, 0, 0, 0, -6, 3, -7])

### 1. How many negative numbers are there?

In [3]:
a[a < 0]

array([-2, -1, -6, -7])

### 2. How many positive numbers are there?

In [4]:
a[a > 0]

array([ 4, 10, 12, 23,  3])

### 3. How many even positive numbers are there?

In [5]:
a[(a > 0) & (a % 2 ==0)]

array([ 4, 10, 12])

### 4. If you were to add 3 to each data point, how many positive numbers would there be?

In [6]:
my_array = a + 3

In [7]:
my_array

array([ 7, 13, 15, 26,  1,  2,  3,  3,  3, -3,  6, -4])

In [8]:
count = my_array[my_array > 0]

In [9]:
len(count)

10

### 5. If you squared each number, what would the new mean and standard deviation be?

In [10]:
squared_array = a **2

In [11]:
squared_array

array([ 16, 100, 144, 529,   4,   1,   0,   0,   0,  36,   9,  49])

In [12]:
std_array = np.std(squared_array)

In [13]:
std_array

144.0243035046516

### 6. A common statistical operation on a dataset is centering. This means to adjust the data such that the mean of the data is 0. This is done by subtracting the mean from each data point. Center the data set. See this link for more on centering.

In [14]:
mean_array = np.mean(a)

In [15]:
mean_array

3.0

In [16]:
center_array = a - 3.0

In [17]:
center_array

array([  1.,   7.,   9.,  20.,  -5.,  -4.,  -3.,  -3.,  -3.,  -9.,   0.,
       -10.])

#### to double check answer

In [18]:
np.mean(center_array)

0.0

### 7.Calculate the z-score for each data point. Recall that the z-score is given by:

```
Z = x - μ
   _______
      σ
```



In [19]:
mean = np.mean(a)

In [20]:
std = np.std(a)

In [21]:
z_score = (a - mean)/(std)

In [22]:
z_score

array([ 0.12403473,  0.86824314,  1.11631261,  2.48069469, -0.62017367,
       -0.49613894, -0.3721042 , -0.3721042 , -0.3721042 , -1.11631261,
        0.        , -1.24034735])

# 8a.

import numpy as np
# Life w/o numpy to life with numpy

### Setup 1
a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

### Use python's built in functionality/operators to determine the following:
### Exercise 1 - Make a variable called sum_of_a to hold the sum of all the numbers in above list

### Exercise 2 - Make a variable named min_of_a to hold the minimum of all the numbers in the above list

### Exercise 3 - Make a variable named max_of_a to hold the max number of all the numbers in the above list

### Exercise 4 - Make a variable named mean_of_a to hold the average of all the numbers in the above list

### Exercise 5 - Make a variable named product_of_a to hold the product of multiplying all the numbers in the above list together

### Exercise 6 - Make a variable named squares_of_a. It should hold each number in a squared like [1, 4, 9, 16, 25...]

### Exercise 7 - Make a variable named odds_in_a. It should hold only the odd numbers

### Exercise 8 - Make a variable named evens_in_a. It should hold only the evens.

### What about life in two dimensions? A list of lists is matrix, a table, a spreadsheet, a chessboard...
### Setup 2: Consider what it would take to find the sum, min, max, average, sum, product, and list of squares for this list of two lists.
b = [
    [3, 4, 5],
    [6, 7, 8]
]




#### 1 - Make a variable called sum_of_a to hold the sum of all the numbers in above list

In [23]:
import numpy as np

In [24]:
a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [25]:
sum_of_a = np.sum(a)

In [26]:
sum_of_a

55

#### or

In [27]:
sum(a)

55

#### 2 - Make a variable named min_of_a to hold the minimum of all the numbers in the above list

In [28]:
min_of_a = np.min(a)

In [29]:
min_of_a

1

#### or

In [30]:
min(a)

1

#### 3 - Make a variable named max_of_a to hold the max number of all the numbers in the above list

In [31]:
max_of_a = np.max(a)

In [32]:
max_of_a

10

#### or

In [33]:
max(a)

10

#### 4 - Make a variable named mean_of_a to hold the average of all the numbers in the above list

In [34]:
mean_of_a = np.mean(a)

In [35]:
mean_of_a

5.5

#### or

In [36]:
sum(a)/len(a)

5.5

#### 5 - Make a variable named product_of_a to hold the product of multiplying all the numbers in the above list together

In [37]:
product_of_a = np.prod(a)

In [38]:
product_of_a

3628800

#### or

In [39]:
product = 1
for item in a:
    product *= item
    print(product)

1
2
6
24
120
720
5040
40320
362880
3628800


#### 6 - Make a variable named squares_of_a. It should hold each number in a squared like [1, 4, 9, 16, 25...]


In [40]:
squares_of_a = np.square(a)

In [41]:
squares_of_a

array([  1,   4,   9,  16,  25,  36,  49,  64,  81, 100])

#### or

In [42]:
for squares in a:
    squares *= squares
    print (squares)

1
4
9
16
25
36
49
64
81
100


#### 7 - Make a variable named odds_in_a. It should hold only the odd numbers¶

In [43]:
import numpy as np

In [44]:
a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [45]:
odds_in_a = [number for number in a if number % 2 != 0]
print(odds_in_a)

[1, 3, 5, 7, 9]


#### 8 - Make a variable named evens_in_a. It should hold only the evens.


In [46]:
evens_in_a = [number for number in a if number % 2 == 0]
print(evens_in_a)

[2, 4, 6, 8, 10]


### What about life in two dimensions? A list of lists is matrix, a table, a spreadsheet, a chessboard...¶


# 8b.

Setup 2: Consider what it would take to find the sum, min, max, average, sum, product, and list of squares for this list of two lists.
b = [ [3, 4, 5], [6, 7, 8] ]

### Exercise 1 - refactor the following to use numpy. Use sum_of_b as the variable. **Hint, you'll first need to make sure that the "b" variable is a numpy array**
```
sum_of_b = 0
for row in b:
    sum_of_b += sum(row)
```
### Exercise 2 - refactor the following to use numpy. 
min_of_b = min(b[0]) if min(b[0]) <= min(b[1]) else min(b[1])  

### Exercise 3 - refactor the following maximum calculation to find the answer with numpy.
max_of_b = max(b[0]) if max(b[0]) >= max(b[1]) else max(b[1])


### Exercise 4 - refactor the following using numpy to find the mean of b
mean_of_b = (sum(b[0]) + sum(b[1])) / (len(b[0]) + len(b[1]))

### Exercise 5 - refactor the following to use numpy for calculating the product of all numbers multiplied together.
```
product_of_b = 1
for row in b:
    for number in row:
        product_of_b *= number
```
### Exercise 6 - refactor the following to use numpy to find the list of squares 
```
squares_of_b = []
for row in b:
    for number in row:
        squares_of_b.append(number**2)
```

### Exercise 7 - refactor using numpy to determine the odds_in_b
```
odds_in_b = []
for row in b:
    for number in row:
        if(number % 2 != 0):
            odds_in_b.append(number)
```


### Exercise 8 - refactor the following to use numpy to filter only the even numbers
```
evens_in_b = []
for row in b:
    for number in row:
        if(number % 2 == 0):
            evens_in_b.append(number)
```

### Exercise 9 - print out the shape of the array b.

### Exercise 10 - transpose the array b.

### Exercise 11 - reshape the array b to be a single list of 6 numbers. (1 x 6)

### Exercise 12 - reshape the array b to be a list of 6 lists, each containing only 1 number (6 x 1)



#### 1 - refactor the following to use numpy. Use sum_of_b as the variable. Hint, you'll first need to make sure that the "b" variable is a numpy array¶
```
sum_of_b = 0
for row in b:
    sum_of_b += sum(row)
```

In [47]:
b = [ [3, 4, 5], [6, 7, 8] ]

In [48]:
sum_of_b = np.sum(b)

In [49]:
sum_of_b

33

#### 2 - refactor the following to use numpy.
min_of_b = min(b[0]) if min(b[0]) <= min(b[1]) else min(b[1])



In [50]:
min_of_b = np.min(b)

In [51]:
min_of_b

3

#### 3 - refactor the following maximum calculation to find the answer with numpy.
max_of_b = max(b[0]) if max(b[0]) >= max(b[1]) else max(b[1])



In [52]:
max_of_b = np.max(b)

In [53]:
max_of_b

8

#### 4 - refactor the following using numpy to find the mean of b¶
mean_of_b = (sum(b[0]) + sum(b[1])) / (len(b[0]) + len(b[1]))

In [54]:
mean_of_b = np.mean(b)

In [55]:
mean_of_b

5.5

#### 5 - refactor the following to use numpy for calculating the product of all numbers multiplied together.
product_of_b = 1
for row in b:
    for number in row:
        product_of_b *= number


In [56]:
product_of_b = np.product(b)

In [57]:
product_of_b

20160

#### 6 - refactor the following to use numpy to find the list of squares
squares_of_b = []
for row in b:
    for number in row:
        squares_of_b.append(number**2)

In [58]:
squares_of_b = np.square(b)

In [59]:
squares_of_b

array([[ 9, 16, 25],
       [36, 49, 64]])

#### 7 - refactor using numpy to determine the odds_in_b
odds_in_b = []
for row in b:
    for number in row:
        if(number % 2 != 0):
            odds_in_b.append(number)

In [60]:
b = np.array([ [3, 4, 5], [6, 7, 8] ])

In [61]:
odds_in_b = b[b % 2 != 0]

In [62]:
odds_in_b

array([3, 5, 7])

#### 8 - refactor the following to use numpy to filter only the even numbers
```
evens_in_b = []
for row in b:
    for number in row:
        if(number % 2 == 0):
            evens_in_b.append(number)
            ```


In [63]:
evens_in_b = b[b % 2 ==0]

In [64]:
evens_in_b

array([4, 6, 8])

#### 9 - print out the shape of the array b.

In [65]:
np.shape(b)

(2, 3)

#### 10 - transpose the array b.

In [66]:
transpose_b = np.transpose(b)

In [67]:
print(transpose_b)

[[3 6]
 [4 7]
 [5 8]]


#### 11 - reshape the array b to be a single list of 6 numbers. (1 x 6)

In [70]:
b = np.array([ [3, 4, 5], [6, 7, 8] ])

reshape = b.reshape(1,6)

In [71]:
reshape

array([[3, 4, 5, 6, 7, 8]])

#### 12 - reshape the array b to be a list of 6 lists, each containing only 1 number (6 x 1)

In [72]:
b = np.array([ [3, 4, 5], [6, 7, 8] ])

reshape = b.reshape(6,1)

In [73]:
reshape

array([[3],
       [4],
       [5],
       [6],
       [7],
       [8]])

# 8c.

### Setup 3
c = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]


## HINT, you'll first need to make sure that the "c" variable is a numpy array prior to using numpy array methods.
### Exercise 1 - Find the min, max, sum, and product of c.
### Exercise 2 - Determine the standard deviation of c.
### Exercise 3 - Determine the variance of c.
### Exercise 4 - Print out the shape of the array c
### Exercise 5 - Transpose c and print out transposed result.
### Exercise 6 - Get the dot product of the array c with c. 
### Exercise 7 - Write the code necessary to sum up the result of c times c transposed. Answer should be 261

### Exercise 8 - Write the code necessary to determine the product of c times c transposed. Answer should be 131681894400.


### Setup 4
d = [
    [90, 30, 45, 0, 120, 180],
    [45, -90, -30, 270, 90, 0],
    [60, 45, -45, 90, -45, 180]
]


# 8d.
### Exercise 1 - Find the sine of all the numbers in d
### Exercise 2 - Find the cosine of all the numbers in d
### Exercise 3 - Find the tangent of all the numbers in d
### Exercise 4 - Find all the negative numbers in d
### Exercise 5 - Find all the positive numbers in d
### Exercise 6 - Return an array of only the unique numbers in d.
### Exercise 7 - Determine how many unique numbers there are in d.
### Exercise 8 - Print out the shape of d.
### Exercise 9 - Transpose and then print out the shape of d.
### Exercise 10 - Reshape d into an array of 9 x 2